iroh-net/bench/src/lib.rs

```diff
@@ -192,3 +142,101 @@ fn parse_byte_size(s: &str) -> Result<u64, ParseIntError> {
 
     Ok(base * multiplier)
 }
+
+#[derive(Default)]
+pub struct ClientStats {
+    upload_stats: Stats,
+    download_stats: Stats,
+    connect_time: std::time::Duration,
+}
+
+impl ClientStats {
+    pub fn print(&self, client_id: usize) {
+        println!();
+        println!("Client {client_id} stats:");
+
+        let ct = self.connect_time.as_nanos() as f64 / 1_000_000.0;
+        println!("Connect time: {ct}ms");
+
+        if self.upload_stats.total_size != 0 {
+            self.upload_stats.print("upload");
+        }
+
+        if self.download_stats.total_size != 0 {
+            self.download_stats.print("download");
+        }
+    }
+}
+
+/// Take the provided endpoint and run the client benchmark
+pub async fn client_handler(
+    endpoint: EndpointSelector,
+    connection: ConnectionSelector,
+    opt: Opt,
+) -> Result<ClientStats> {
+    let start = Instant::now();
+
+    let connection = Arc::new(connection);
+
+    let mut stats = ClientStats::default();
+    let mut first_error = None;
+
+    let sem = Arc::new(Semaphore::new(opt.max_streams));
+    let results = Arc::new(Mutex::new(Vec::new()));
+    for _ in 0..opt.streams {
+        let permit = sem.clone().acquire_owned().await.unwrap();
+        let results = results.clone();
+        let connection = connection.clone();
+        tokio::spawn(async move {
+            let result = match &*connection {
+                ConnectionSelector::Iroh(connection) => {
+                    iroh::handle_client_stream(connection, opt.upload_size, opt.read_unordered)
+                        .await
+                }
+                ConnectionSelector::Quinn(connection) => {
+                    quinn::handle_client_stream(connection, opt.upload_size, opt.read_unordered)
+                        .await
+                }
+            };
+            // handle_client_stream(connection, opt.upload_size, opt.read_unordered).await;
+            info!("stream finished: {:?}", result);
+            results.lock().unwrap().push(result);
+            drop(permit);
+        });
+    }
+
+    // Wait for remaining streams to finish
+    let _ = sem.acquire_many(opt.max_streams as u32).await.unwrap();
+
+    for result in results.lock().unwrap().drain(..) {
+        match result {
+            Ok((upload_result, download_result)) => {
+                stats.upload_stats.stream_finished(upload_result);
+                stats.download_stats.stream_finished(download_result);
+            }
+            Err(e) => {
+                if first_error.is_none() {
+                    first_error = Some(e);
+                }
+            }
+        }
+    }
+
+    stats.upload_stats.total_duration = start.elapsed();
+    stats.download_stats.total_duration = start.elapsed();
```

Comment:
maybe move this up to right after `acquire_many()`?  Probably not a big difference though